File: 0003_Longest_Substring_Without_Repeating_Characters/solution.ts
function lengthOfLongestSubstring(s: string): number {
    let arr = Array<number>(128).fill(0);
    let start = 0;
    let max = 0;

    for (let end = 0; end < s.length; end++) {
        let endCode = s.charCodeAt(end);
        while (arr[endCode] > 0) {
            let startCode = s.charCodeAt(start);
            arr[startCode]--;
            start++;
        }
        arr[endCode]++;
        max = Math.max(max, end - start + 1);
    }
    return max;
}

File: 0003_Longest_Substring_Without_Repeating_Characters/solution.c
#include <math.h>


int length_of_longest_substring(char *s) {
    int freq[128] = {0};
    int start = 0;
    int end = 0;
    int max = 0;

    while (s[end] != '\0') {
        if (++freq[s[end]] < 2) {
            max = fmax(max, end - start + 1);
        } else {
            while (freq[s[end]] > 1) {
                --freq[s[start++]];
            }
        }
        end++;
    }
    return max;
}